yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

```diff
@@ -372,7 +372,14 @@ private[spark] class ApplicationMaster(
             case i: InterruptedException =>
             case e: Throwable => {
               failureCount += 1
-              if (!NonFatal(e) || failureCount >= reporterMaxFailures) {
+              if ("org.apache.hadoop.yarn.exceptions.ApplicationAttemptNotFoundException".equals(
+                e.getClass().getName())) {
+                val message = "ApplicationAttemptNotFoundException was thrown " +
+                  "from Reporter thread."
+                logError(message, e)
+                finish(FinalApplicationStatus.FAILED, ApplicationMaster.EXIT_REPORTER_FAILURE,
+                  message)
+              } else if (!NonFatal(e) || failureCount >= reporterMaxFailures) {
```

Comment:
If the exception is fatal then it would finish the application master immediately without continuing retries, otherwise it will retry with the allocate. I think throwing the fatal exception without catching or finishing the application master on fatal exception wouldn’t make big difference here.